### PR TITLE
decoder: Implement completion for binary & unary operators

### PR DESCRIPTION
**Pull request by radeksimko**
Related:
 - https://github.com/hashicorp/terraform-ls/issues/529

Follow-up issue:
 - https://github.com/hashicorp/hcl-lang/issues/321
 - https://github.com/hashicorp/hcl-lang/issues/325

---

### CODE REVIEW

**Review by dbanck — COMMENTED** [review]
LGTM! 👍 

I'm not sure how extensively we want to test this. Since any expressions deal with references, functions and literal types, should we cover all three in all new tests?

File: decoder/expr_any_completion_test.go [hunk 1]
```diff
@@ -2826,3 +2826,393 @@ func TestCompletionAtPos_exprAny_skipComplex(t *testing.T) {
...
+			`attr = 42 + var.
+`,
+			hcl.Pos{Line: 1, Column: 17, Byte: 16},
+			lang.CompleteCandidates([]lang.Candidate{
+				// TODO: See https://github.com/hashicorp/hcl-lang/issues/321
+			}),
+		},
+	}
+
+	for i, tc := range testCases {
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply decoder/expr_any_completion_test.go:3192]: Yeah, I think that's a good idea. Initially I thought that all we should need to be testing is that the "hand-over" happens between expressions but in reality I reckon there may be some other edge cases with "special" characters like `(` or `)` when functions are involved and that may be worth testing.

[the remark this answers is not in the document]
Comment (radeksimko) [reply decoder/expr_any_completion_test.go:3400]: I added two more tests for functions and filed an issue for a related edge case in https://github.com/hashicorp/hcl-lang/issues/325

**Comment by radeksimko** [comment]
> I'm not sure how extensively we want to test this. Since any expressions deal with references, functions and literal types, should we cover all three in all new tests?

I would argue that some tests with functions are useful, as mentioned - and as added 😄 - specifically in completion, because incomplete configuration just brings so many more edge cases. 😭 

However I'm not sure we'd actually gain much by adding tests to hover, semantic tokens or origins. The logic we have in place outside of completion should just treat any/all nested expression types the same (it's all `hcl.Expression` on the surface) and so one test with _any_ expression type (such as traversal) should be sufficient to test _that_.

**Review by dbanck — APPROVED** [review]
Thank you for adding the tests
